assembly/src/parsers/field_ops.rs

```diff
@@ -373,6 +383,54 @@ pub fn parse_gte(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Assemb
     Ok(())
 }
 
+// POWER OF TWO HELPER FUNCTIONS
+// ================================================================================================
+
+/// Splits the top element on the stack into its binary decomposition and performs a
+/// power of 2 bitwise accumulation.
+/// The expected starting state of the stack (from the top) is: [a, ...].
+fn aggregate_power_2(span_ops: &mut Vec<Operation>, unsafe_mode: bool) {
+    const MOST_SIGNIFICANT_BIT: u32 = 5;
+    let base = 2_u64;
+
+    // `safe` Mode
+    if !unsafe_mode {
+        // Checks if the top element of the stack is less than 64 or not.
+        span_ops.push(Operation::Dup0);
+        span_ops.push(Operation::Push(Felt::new(64)));
+        check_lt(span_ops);
+        span_ops.push(Operation::Assert);
+    }
+
+    // Decompose the number to its binary representation. Eg. 59 will be represent by
+    // 1100111 (most significant bit on the right most side or deeper in the stack).
+    for idx in (1..=MOST_SIGNIFICANT_BIT).rev() {
+        span_ops.extend_from_slice(&[Operation::Push(Felt::new(base.pow(idx))), Operation::U32div]);
+    }
+
+    // aggregate the bits from the binary representation to form power of 2 of the
+    // original input. 2^59 can be aggregated using: ((2^0 - 1)*1 + 1) * ((2^1 - 1)*1 + 1)
+    // * ((2^2 - 1)*0 + 1) * ((2^3 - 1)*0 + 1) * ((2^4 - 1)*1 + 1) * ((2^5 - 1)*1 + 1).
+    span_ops.push(Operation::Pad);
+    span_ops.push(Operation::Incr);
+    span_ops.push(Operation::Swap);
+    span_ops.push(Operation::Incr);
+    span_ops.push(Operation::Mul);
+    span_ops.push(Operation::Swap);
+
+    for idx in 1..=MOST_SIGNIFICANT_BIT {
+        let exponent = base.pow(idx) as u32;
+        let exponent_of_two = base.pow(exponent);
+        span_ops.push(Operation::Push(Felt::new(exponent_of_two - 1)));
+        span_ops.push(Operation::Mul);
+        span_ops.push(Operation::Incr);
+        span_ops.push(Operation::Mul);
+        if idx != MOST_SIGNIFICANT_BIT {
+            span_ops.push(Operation::Swap);
+        }
+    }
+}
```

Comment:
I think we might be able to do the unsafe version in 36 operations. I've put together the sequence below, could you check if it works?

```
push.2
u32div.unsafe
incr
swap
push.2
u32div.unsafe
push.3
mul
incr
swap
push.2
u32div.unsafe
push.15
mul
incr
swap
push.2
u32div.unsafe
push.(2^8 -1)
mul
incr
swap
push.2
u32div.unsafe
push.(2^16-1)
mul
incr
swap
push.(2^32-1)
mul
incr
mul
mul
mul
mul
mul
```
If the above works, we could also eliminate most of `push.2` by adding 2 extra cycles up front.

Reply:
Nice, It's working! 

> If the above works, we could also eliminate most of push.2 by adding 2 extra cycles up front.

By the above, do you mean we should insert `push.2` & `swap` operations upfront, and every time whenever we need a `push.2` operation we can do `dupn` (we will know deterministically where is our 2 in the stack) instead? If this is the case then shouldn't we remove this 2, which was initially added?

Reply:
Great! And nice sheet! (though, if you had operations listed vertically, it would be a bit easier to follow).



> By the above, do you mean we should insert `push.2` & `swap` operations upfront, and every time whenever we need a `push.2` operation we can do `dupn` (we will know deterministically where is our 2 in the stack) instead? If this is the case then shouldn't we remove this 2, which was initially added?

Yes, exactly! We don't need to remove the 2 at the end because the last time we need it we can use `movup` instead of `dup`.

Reply:
I've updated the sheet as per your suggestions. 

> Yes, exactly! We don't need to remove the 2 at the end because the last time we need it we can use movup instead of dup.

Makes sense, thanks! Please see [7a4b5dc](https://github.com/maticnetwork/miden/pull/284/commits/7a4b5dcea51606cbaff2a203a367f24f5215a75d).